### PR TITLE
Better pattern matcher error message.

### PR DESCRIPTION
**Pull request by paulp**
For the common case when someone hasn't quite grokked
the significance of lower case in a pattern match.  I'd
like to make all the unreachables errors, not warnings,
but there may be a bug or two to clear out first.

```
class A {
  def badEquals(x: Any, y: Any) = x match {
    case y => true
    case _ => false
  }
}
```

a.scala:3: warning: patterns after a variable pattern cannot match (SLS 8.1.1)
If you intended to match against parameter y of method badEquals, you must use backticks, like: case `y` =>
    case y => true
         ^
a.scala:4: warning: unreachable code due to variable pattern 'y' on line 3
    case _ => false
              ^
two warnings found

### CODE REVIEW

**Comment by paulp** [comment]
Review by @adriaanm

**Comment by adriaanm** [comment]
my last two remaining critical bugs are related to unreachability: SI-6146, SI-6022
i'll be spending the remainder of my working days of this summer on them

**Comment by paulp** [comment]
The logic might move around, but this pull request should be useful regardless; it's more about the error message.

**Comment by adriaanm** [comment]
> but there may be a bug or two to clear out first.
> ^^^ this was the extent of what I was replying to

**Comment by paulp** [comment]
Oh, I forgot I wrote that.

**Comment by paulp** [comment]
Thanks random spurious failure.

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
ah the mysteries of spuriousness

```
BUILD SUCCESSFUL
Total time: 65 minutes 17 seconds
+ buildState=0
+ /localhome/hudson/b/workspace/pr-scala-testsuite-linux-opt/jenkins-scripts/archive-checkin
++ git rev-parse HEAD
+ revision=9213c7d6a3fef728b065e69ffb4f646763200e49
+ rm -rf jenkins-artifacts
+ mkdir jenkins-artifacts
+ antWithMavenRev=81985e49cf9f492a4af6be8ed5c3125ca4861fbe
+ antNoMavenRev=0385e9835d6fac729b6672294fa8e4c6b1ea3c10
++ git rev-list 9213c7d6a3fef728b065e69ffb4f646763200e49
++ grep 81985e49cf9f492a4af6be8ed5c3125ca4861fbe
+ [[ -n 81985e49cf9f492a4af6be8ed5c3125ca4861fbe ]]
+ antWithMaven
+ echo 'archiving antWithMaven'
archiving antWithMaven
+ buildPack
+ [[ -d build/pack ]]
+ tar czf jenkins-artifacts/pack.tgz -C build pack
+ distsLatest
+ [[ -L dists/latest ]]
+ tar czhf jenkins-artifacts/dist.tgz -C dists latest
+ distsMavenLatest
+ [[ -d dists/maven/latest ]]
+ tar czhf jenkins-artifacts/maven.tgz -C dists/maven latest
+ [[ '' == \p\u\b\l\i\s\h ]]
+ publishState=1
+ [[ 0 == 0 ]]
+ [[ 1 == 0 ]]
+ exit 1
Build step 'Execute shell' marked build as failure
Archiving artifacts
Finished: FAILURE
```

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -198,6 +198,69 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
     import typer.{typed, context, silent, reallyExists}
     // import typer.infer.containsUnchecked
 
+    // Why is it so difficult to say "here's a name and a context, give me any
+    // matching symbol in scope" ? I am sure this code is wrong, but attempts to
+    // use the scopes of the contexts in the enclosing context chain discover
+    // nothing. How to associate a name with a symbol would would be a wonderful
+    // linkage for which to establish a canonical acquisition mechanism.
+    def matchingSymbolInScope(pat: Tree): Symbol = {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:206]: the slow but correct  way would be to call typedIdent on a typer in that context, i guess

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:206]: though that would have to account for ambiguities etc

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**